backend/server/acl_test_helper.go

```diff
@@ -169,3 +171,63 @@ var isMemberOfAnyProjectOwnsDatabase = func(principalID int, databaseID int) (bo
 	}
 	return false, nil
 }
+
+var testMemberIssueMap = struct {
+	// principalIDToProjectID is a map from principal ID to the project ID.
+	// We assume that a principal can only be a member of one project (actually it can be a member of multiple projects).
+	principalIDToProjectID map[int]int
+	// issueIDToProjectID is a map from issue ID to the project ID.
+	issueIDToProjectID map[int]int
+	// issueIDToCreatorID is a map from issue ID to the creator ID.
+	issueIDToCreatorID map[int]int
+}{
+	principalIDToProjectID: map[int]int{
+		// User 202 is a member of project 102.
+		202: 102,
+		// User 203 is a member of project 102.
+		203: 102,
+		// User 203 is a member of project 103.
+		204: 103,
+	},
+	issueIDToProjectID: map[int]int{
+		// Issue 401 belongs to project 102.
+		401: 102,
+		// Issue 402 belongs to project 102.
+		402: 102,
+		// Issue 403 belongs to project 103.
+		403: 103,
+	},
+	issueIDToCreatorID: map[int]int{
+		// Issue 401 is created by user 202.
+		401: 202,
+		// Issue 402 is created by user 203.
+		402: 203,
+		// Issue 403 is created by user 204.
+		403: 204,
+	},
+}
+
+// If the workspace developer principal is one of the following, it can update the issue:
+// 1. The creator of the issue.
+// 2. The member of the project that the issue belongs to.
+var canWorkspaceDeveloperUpdateIssue = func(issueID int, principalID int) error {
+	if creatorID, ok := testMemberIssueMap.issueIDToCreatorID[issueID]; !ok {
+		return errors.Errorf("issue %d does not exist", issueID)
+	} else if creatorID == principalID {
```

Comment:
my feeling is we should write a callback to find the project member, instead of a more complex canWorkspaceDeveloperUpdateIssue.

Currently, we are almost overwriting the implementation entirely, which makes the test kind of meaningless.

Reply:
Good idea, we should mock the store layer instead of the nearly full logic.

Reply:
I split them into two units - getIssueCreatorAndProjectID and getIssueMembers, but IMO it's still complex, I think we should migrate to using integration test to test this logic when we launch API V2.